### PR TITLE
Implement an improved Google Cloud Function for BT automation

### PR DESCRIPTION
**Pull request by pradh**
Details on what it does and how to test it in README.md.

Features, compared to current flow:
- Directly interact with the Dataflow job, cutting an additional Cloud Function and Airflow system from the loop.
- Make synchronization per-cache (instead of singleton) so we can run concurrent builds.
- Handle both base and branch caches (concurrently too).
- Allow building locally and validating in test environment. Adding an integration test is pretty painful given the many pieces, but this should be good for now.

I've validated the flow in a branch cache in test environment.

Next steps: change Google pipeline to sync on the new paths, and test for reals...

---
This change is [Reviewable](https://reviewable.io/reviews/datacommonsorg/tools/84)

### CODE REVIEW

**Review by hanlu09205 — COMMENTED** [review]
Nice work!

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/tools/84#-:-MQuDKfW0eDCH7a26xe3:bdtx8r6)* status: all files reviewed, 3 unresolved discussions (waiting on @pradh and @shifucun)

___
*[cloud_automation/gcf/bt_trigger.go, line 29 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuBwKlBCFTvAezGQxi:-MQuBwKlBCFTvAezGQxj:bu5c6et) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/bt_trigger.go#L29)):*
> ```Go
> 
> const (
> 	PROD							= "prod"
> ```

Nit, is there a way for format this go file?
___
*[cloud_automation/gcf/go.mod, line 1 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuCEcoAoG6Gus_ypxQ:-MQuCEcoAoG6Gus_ypxR:bksu2ic) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/go.mod#L1)):*
> ```XML
> module btcachegeneration
> ```

For my education, is this some kind of configuration file?

And what about go.sum?
___
*[cloud_automation/gcf/README.md, line 32 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuCjom7sy8dO4O7rmn:-MQuCjom7sy8dO4O7rmo:b-yum9bs) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/README.md#L32)):*
> ```Markdown
> ```
> 
> ## Validate in Test Environment
> ```

This is for validation the tool?
And the test environment is BT test environment?

**Review by shifucun — COMMENTED** [review]
Can try https://blog.golang.org/gofmt to fmt the cods

File: cloud_automation/gcf/go.sum [hunk 1]
```diff
@@ -0,0 +1,461 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (shifucun) [inline cloud_automation/gcf/go.sum:1]: the .sum file seems to be too long given the go.mod file only have a few library

File: cloud_automation/gcf/bt_trigger.go [hunk 2]
```diff
@@ -20,31 +20,94 @@ import (
...
+	// TODO: Deprecate these files.
+	triggerFile				= "latest_base_cache_version.txt"
+	successFile				= "success.txt"
+	failureFile				= "failure.txt"
 	airflowTriggerFile = "airflow_trigger.txt"
 )
 
+type Environment struct {
+	// BT instance holding base caches.
+	BaseBTInstance string
```
Comment (shifucun) [inline cloud_automation/gcf/bt_trigger.go:62]: nit: Capital names means it's exported and used by other modules, if it's internal can just be lower case

**Review by shifucun — APPROVED** [review]
nice work, good this can be tested locally

**Comment by shifucun** [comment]
Can also try https://github.com/golang/lint, Go does very good job to cache common issues from lint tools and standard practice

**Review by pradh — COMMENTED** [review]
Thanks for the gofmt and golint reminders. Ran them..

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/tools/84#-:-MQwVh2vDWqNiYdFEZ3b:bkc5kgn)* status: 4 of 6 files reviewed, 5 unresolved discussions (waiting on @hanlu09205 and @shifucun)

___
*[cloud_automation/gcf/bt_trigger.go, line 29 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuBwKlBCFTvAezGQxi:-MQwRpaxDnyUaVxkeKk9:b-rda4v5) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/bt_trigger.go#L29)):*
<details><summary><i>Previously, hanlu09205 wrote…</i></summary><blockquote>

Nit, is there a way for format this go file?
</blockquote></details>

Ran gofmt
___
*[cloud_automation/gcf/bt_trigger.go, line 62 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuBtqphOA4Q348Wsnh-r1-62:-MQwS7zcAjC7lBnPAZgC:bzbbn8x) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/bt_trigger.go#L62)):*
<details><summary><i>Previously, shifucun (Bo Xu) wrote…</i></summary><blockquote>

nit: Capital names means it's exported and used by other modules, if it's internal can just be lower case
</blockquote></details>

Ah, thanks... fixed it, and passes golint.
___
*[cloud_automation/gcf/go.mod, line 1 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuCEcoAoG6Gus_ypxQ:-MQwRVfI66IiyKtd0w18:b-gbz9hs) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/go.mod#L1)):*
<details><summary><i>Previously, hanlu09205 wrote…</i></summary><blockquote>

For my education, is this some kind of configuration file?

And what about go.sum?
</blockquote></details>

New to me too. Got auto-generated when running `go mod init`
___
*[cloud_automation/gcf/go.sum, line 1 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuBtqphOA4Q348Wsnl-r1-1:-MQwVXkLCq0BPuyaWdDT:bwlssrf) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/go.sum#L1)):*
<details><summary><i>Previously, shifucun (Bo Xu) wrote…</i></summary><blockquote>

the .sum file seems to be too long given the go.mod file only have a few library
</blockquote></details>

It happens that way after I run the main.go...
___
*[cloud_automation/gcf/README.md, line 32 at r1](https://reviewable.io/reviews/datacommonsorg/tools/84#-MQuCjom7sy8dO4O7rmn:-MQwR6mo8cTZRoB8dKOo:b-fiph2i) ([raw file](https://github.com/datacommonsorg/tools/blob/364140d3e50701185f9d1c049d97b1760b926eb9/cloud_automation/gcf/README.md#L32)):*
<details><summary><i>Previously, hanlu09205 wrote…</i></summary><blockquote>

This is for validation the tool?
And the test environment is BT test environment?
</blockquote></details>

Yes, clarified heading more...

**Review by shifucun — APPROVED** [review]

**Comment by pradh** [comment]
Thanks!

**Pull request merged**